### ISSUE TITLE
Support of additional token endpoint auth methods

### ISSUE THREAD

**Issue by oddsome** [pjones/openid-connect#3]
Hello! It looks like the "discovery" request fails when the discovery response contains "token_endpoint_auth_methods_supported" which aren't explicitly listed in https://github.com/pjones/openid-connect/blob/trunk/src/OpenID/Connect/Authentication.hs#L161
Unfortunately, such cases aren't uncommon.

There are a few things to note here:
1. Spec (https://www.rfc-editor.org/rfc/rfc8705.html#name-pki-method-metadata-value) allows "tls_client_auth" as "token_endpoint_auth_methods_supported" but since it's not explicitly listed as legit "ClientAuthentication" - openid-connect can't work with this.
2. Spec allows additional client authentication methods using extensions (https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata)

In summary, perhaps the best solution would be to add a constructor like "Other Text" for "ClientAuthentication", or just skip unfamiliar authentication methods in presence of legit ones. 
It should also be explicitly stated that openid-connect works with a specific subset of "token_endpoint_auth_methods_supported" because of "tls_client_auth".

I would like to help solve this issue, because it prevents me from using openid-connect, so any feedback or advice would be appreciated, thanks.

**Comment by pjones** [comment]
Hey there @oddsome.

I think the correct solution is a new type:

```haskell
data TokenEndpointAuthMethod
  = StandardAuthentication ClientAuthentication
  | AuthenticationExtension Text
```

And use the new type in the discovery record.

That would allow the existing code (namely `applyRequestAuthentication`) to remain type safe since there's no way for it to know how to authenticate using an extension.   It would then be the responsibility of the library user to craft an appropriate authentication technique for any given extension.

If there's a particularly popular extension we could include a function in this package that would authenticate with it.  Maybe something like this:

```haskell
data AuthExtension
  = FooExtension
  | BarExtension

authenticationExtension :: Text -> Maybe AuthExtension

applyRequestAuthUsingExtension :: AuthExtension -> ...
```

Are you willing to submit a patch/PR?

**Comment by oddsome** [comment]
I will look at everything once again, and then I will try to fix the issue and send a pull request

**Comment by pjones** [comment]
@oddsome Fantastic.  If there's anything you need please let me know.
